src/bulk/common.ts

```diff
@@ -1,43 +1,195 @@
   isPromiseLike,
   hasAtomicOperators,
   maxWireVersion,
-  Callback
+  Callback,
+  MongoDBNamespace
 } from '../utils';
+import { PromiseProvider } from '../promise_provider';
+import { Long, ObjectId, Document, BSONSerializeOptions } from '../bson';
+import { MongoError, MongoWriteConcernError } from '../error';
 import { executeOperation } from '../operations/execute_operation';
 import { InsertOperation } from '../operations/insert';
 import { UpdateOperation } from '../operations/update';
 import { DeleteOperation } from '../operations/delete';
-import type { WriteConcern } from '../write_concern';
+import { WriteConcern, WriteConcernOptions } from '../write_concern';
+import type { HasRetryableWrites } from './../utils';
+import type { Collection } from '../collection';
+import type { Topology } from '../sdam/topology';
+import type { Hint } from './../operations/operation';
+import type { addToOperationsList as orderedAddToOperationsList } from './ordered';
+import type { addToOperationsList as unorderedAddToOperationsList } from './unordered';
+import type { CollationOptions } from '../cmap/wire_protocol/write_command';
+import type { InsertOptions } from './../operations/insert';
+import type { RemoveOptions } from './../cmap/wire_protocol/index';
+import type { ReplaceOptions } from './../operations/replace_one';
+import type { UpdateOptions } from './../operations/update';
+
+interface FinalOptionsConfigOptions {
+  bypassDocumentValidation?: boolean;
+  retryWrites?: boolean;
+}
+interface ResultHandler {
+  (err?: AnyError, result?: any): void;
+  operationId?: number;
+}
 
 // Error codes
 const WRITE_CONCERN_ERROR = 64;
 
-// Insert types
-const INSERT = 1;
-const UPDATE = 2;
-const REMOVE = 3;
+export enum BatchType {
+  INSERT = 1,
+  UPDATE = 2,
+  REMOVE = 3
+}
+
+/** @public */
+export interface Operation {
+  selector?: Document;
+  multi?: boolean;
+  upsert?: boolean;
+  limit?: number;
+}
+
+/** @public */
+export interface InsertOneModel extends Document, InsertOptions {
+  document?: Document;
+  _id?: ObjectId;
+}
+
+/** @public */
+export type InsertManyModel = InsertOneModel[];
+
+/** @public */
+export interface ReplaceOneModel extends ReplaceOptions {
+  q?: Document;
+  replacement: Document;
+  filter?: Document;
+}
+
+/** @public */
+export interface UpdateOneModel extends UpdateOptions {
+  q?: Document;
+  update: Document;
+  filter?: Document;
+}
+
+/** @public */
+export interface UpdateManyModel extends UpdateOptions {
+  q?: Document;
+  update: Document;
+  filter?: Document;
+}
+
+/** @public */
+export type RemoveOneModel = RemoveOptions;
+
+/** @public */
+export interface RemoveManyModel extends RemoveOptions {
+  limit?: number;
+}
+
+/** @public */
+export interface DeleteOneModel extends RemoveOptions {
+  q?: Document;
+}
+
+/** @public */
+export interface DeleteManyModel extends RemoveOptions {
+  q?: Document;
+}
+
+/** @public */
+export interface BatchOperations {
```

Comment:
I thought we discussed not using a partial approach like this, but instead using a type hierarchy to represent the models:
```
interface WriteModel {}
interface InsertModel extends WriteModel {}
interface UpdateModel extends WriteModel {}
type AnyWriteModel = InsertModel | UpdateModel | ...;
class BulkOperationBase {
  ...
  insert(model: InsertModel);
  update(model: UpdateModel);
  raw(op: AnyWriteModel);
  ...
}
```

Was there something that was preventing this approach technically?